### PR TITLE
Recurrent Neural Networks

### PR DESCRIPTION
**Pull request by sshekh**
I am a GSoC student working on introducing Recurrent Neural Networks in TMVA. This is the rectified commits using suggestions from #693 . It introduces basic `RNNLayer`, `RecurrentNet` with tests for forward pass. It does not require immediate merge and I'll keep adding commits to it according to feedbacks.

### CODE REVIEW

File: tmva/tmva/test/DNN/RNN/TestForwardPass.cxx [hunk 1]
```diff
@@ -2,15 +2,15 @@
...
- * Copyright (C) 2017, Saurav Shekhar                                    *
- * All rights reserved.                                                  *
- *                                                                       *
- * For the licensing terms see $ROOTSYS/LICENSE.                         *
- * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ * Copyright (C) 2017, Saurav Shekhar                                                      *
+ * All rights reserved.                                                                          *
+ *                                                                                                        *
+ * For the licensing terms see $ROOTSYS/LICENSE.                                     *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.                   *
```
Comment (amadio) [inline tmva/tmva/test/DNN/RNN/TestForwardPass.cxx]: The changes above are quite strange. It looks like this is only misaligning the stars. This happens elsewhere too. Please drop this and similar parts of this commit (where blanks misalign the comments/code).

**Comment by amadio** [comment]
Thank you for the updated PR. Since you are introducing mostly new code, it would be nice to format everything with clang-format and amend the commit, so that it's in the proper format of our coding standards when it's added to git. Please let us know if you have any issues with that and we can help you.

**Comment by amadio** [comment]
@phsft-bot build

**Comment by sshekh** [comment]
@amadio Thank you for the feedback. I will fix the indentation issue. The reason for misalignment of `*` was that I had written the original code with 2 spaces of indentation and later changed them all to 3. Regarding the `clang-format-diff`. I tried to get it working, but it does not seem to work. Especially, it is unable to capture filenames correctly from `git diff`. Details can be seen in the bug filed with clang here [https://bugs.llvm.org/show_bug.cgi?id=33759](https://bugs.llvm.org/show_bug.cgi?id=33759)

**Comment by bluehood** [comment]
Hi, instead of clang-ormat-diff just run clang-format on the new files, that will work.

**Comment by amadio** [comment]
Also, please amend the previous commit instead of adding a commit with only whitespace changes to a file that was just added to git. Thanks.

**Comment by omazapa** [comment]
Hi Saurav
@sshekh do you have this code in the pull 817 with Vladimir and Akshay?

Cheers,
Omar.

**Comment by sshekh** [comment]
@omazapa yes the code is part of the Deep Learning module PR. I will close this

**Pull request closed without merging**